Fix installation of a tag from repos with many tags

alexcrichton opened the pull request:
Attempting to fix https://github.com/bytecodealliance/actions/issues/14

alexcrichton committed:
Update versions of actions used

rajatjindal commented:
Hi @alexcrichton, aplogies i missed this issue earlier. just got notified again. I can check today itself so that you can focus on other priority stuff.

alexcrichton commented:
Oh no worries! I figured I'd give things a shot and I think I'm close to a fix, just wanted to see some failures on CI first which are now rolling in, so now I get to see if my fix fixes those failures...

alexcrichton committed:
Fix installing older tags

Don't use `listReleases` but instead use methods which look like they're
targeted for a particular release.

rajatjindal commented:
> Oh no worries! I figured I'd give things a shot and I think I'm close to a fix, just wanted to see some failures on CI first which are now rolling in, so now I get to see if my fix fixes those failures...

thank you.

alexcrichton commented:
Ok looks like that fixed ci, yay! CI is failing currently in this repo due to some actions using historical/deprecated versions and additionally due to #14, so this PR is effectively fixing CI again at the same time as fixing #14.

@rajatjindal if you're able to review that'd be most welcome, but if you're busy as well I'm happy to work on merging/tagging this too!

alexcrichton commented:
(also no rush of course, feel free to get to this at your leisure!)

rajatjindal commented:
Hi @alexcrichton, thanks for this PR. 

I think we also need to fix `getDownloadLink` to handle pagination. and IIRC there was a scenario where getting the latest release may not work (e.g. if it is a monorepo and we publish multiple release-artifacts from the same repo), which is why I had used listReleases function (but I couldn't find a example repo for this during my brief search).

does that make sense? if so, I can submit the PR that makes use of [pagination](https://docs.github.com/en/rest/using-the-rest-api/using-pagination-in-the-rest-api?apiVersion=2022-11-28#example-using-the-octokitjs-pagination-method) and fixes this issue. But also happy to continue to review this PR if you think otherwise.

alexcrichton committed:
Restore old behavior of `getLatestRelease`, but use pagination

alexcrichton commented:
Sounds good to me, but to confirm you meant the `getLatestRelease` function right? I've left` getDownloadLink` as fetching a release directly, but I've restored `getLatestRelease` to the old behavior (plus new pagination support

rajatjindal commented:
> Sounds good to me, but to confirm you meant the `getLatestRelease` function right? I've left` getDownloadLink` as fetching a release directly, but I've restored `getLatestRelease` to the old behavior (plus new pagination support)

I meant to say we may need pagination in both. because in the `getDownloadLink` we are getting all releases (again) and finding the release corresponding to the tag, so we may endup with a scenario where the release is on next page.

alexcrichton commented:
Oh hm I may be a bit confused then. In `getLatestRelease` it makes sense to use pagination iterating the releases, but `getDownloadLink` was updated to use `getReleaseByTag` instead of `listReleases` which I didn't think required any pagination. Can you highlight which part needs updating in `getDownloadLink`?

rajatjindal commented:
> Oh hm I may be a bit confused then. In `getLatestRelease` it makes sense to use pagination iterating the releases, but `getDownloadLink` was updated to use `getReleaseByTag` instead of `listReleases` which I didn't think required any pagination. Can you highlight which part needs updating in `getDownloadLink`?

ah, I missed that change in the PR. yeah, that should be good IMO. 

Thanks Alex.
